### PR TITLE
Silence "Ambiguous Scalar" Warnings

### PR DESCRIPTION
**Pull request by DabeDotCom**
### Summary
Silence `Warning: Use of "scalar" without parentheses is ambiguous at template path/to/template.html.ep line XXX` messages

### Motivation
I included a module that (like Mojolicious itself) enabled `use warnings` in its calling context.  Unfortunately, however, whenever that template included an `%= expression` (or `<%= expression %>`) it would issue a warning to STDERR.

I had originally looked into wrapping the "scalar + $value" like `scalar( + $value )`, and/or `scalar do { + $value }`, but that didn't work when the expression spanned multiple lines.  :-/

Then I tried adding just the `do {` to the opening statement, and the `}` before the semicolon, but *that* didn't work for, e.g., `%= $block->(--$i) if $i` — so I ended up just punting and prepending `no warnings 'ambiguous';` before the whole assignment.

"If it's stupid — but it works — it isn't stupid..."  ;-D

### CODE REVIEW

**Comment by Grinnz** [comment]
I don't think this is the best fix either, this warning is important for many instances which the user might run into themselves.

**Comment by Grinnz** [comment]
What version of Perl are you seeing these warnings on?

**Comment by DabeDotCom** [comment]
> this warning is important for many instances which the user might run into themselves

I do believe, however, that it's limited in scope to just the template, at least...  (Though there might be a better way to disable the warning once at the beginning of the whole `$_O` construction, rather than every time an expression is added.)

> What version of Perl are you seeing these warnings on?

I just tested it w/ Mojolicious 8.12 under Perl 5.24.0 (though I first noticed it under 5.18.)

`perl -MMojo::Template -E 'say Mojo::Template->new->render(q{<% use warnings; %><%= "ABC" %>})'`

**Comment by jhthorsen** [comment]
Why can't we just do this?

```
@@ -213,11 +213,11 @@ sub _compile {
 
       # Escaped
       if (!$multi && ($op eq 'escp' && !$escape || $op eq 'expr' && $escape)) {
-        $blocks[-1] .= "\$_O .= _escape scalar + $value";
+        $blocks[-1] .= "\$_O .= _escape scalar(+ $value)";
       }
 
       # Raw
-      elsif (!$multi) { $blocks[-1] .= "\$_O .= scalar + $value" }
+      elsif (!$multi) { $blocks[-1] .= "\$_O .= scalar(+ $value)" }
```

**Comment by DabeDotCom** [comment]
@jhthorsen  If *only* it were that easy!  ;-D

Like I said in my original message, there's no guarantee that the entire statement will be self-contained in a single <%= template %> expression...  It could just be part of a larger "begin/end" block, or it could have trailing conditionals, etc.

For example:

**`print Mojo::Template->new->render("<% use warnings; %>Two + Two = <%= 4 if 0 %>");`**

gets compiled to:

```
use warnings; $_O .= "Two\ \+\ Two\ \=\ ";$_O .= scalar +  4 if 0 ;$_O .= "\n";
```

But putting parentheses around `scalar(+ $value)` turns that into:

```
use warnings; $_O .= "Two\ \+\ Two\ \=\ ";$_O .= scalar(+  4 if 0 );$_O .= "\n";
```

which is a syntax error; the `if 0` needs to be OUTSIDE the parens.

Like I said, I also tried using, like, `do { $value }` and stuff like that, instead, but things still went haywire if the template spanned multiple lines:

```
<%= join " " =>
    $user->{'firstname'},
    $user->{'lastname'},
    $user->{'age'},
%>
```

etc.

Conveniently, `t/mojo/template.t` tests a whole slew of edge cases [Yay TDD!], and so far, my `no warnings 'ambiguous'` hack^H^H^H^H _solution_ is the only way I've found to satisfy them all...  :-}

**Comment by kraih** [comment]
There is just no way we prepend `no warnings 'ambiguous';` to every expression.

**Comment by jhthorsen** [comment]
@DabeDotCom: Sorry for my idiotic comment. I have no idea how I could run my suggested patch and _not_ notice the failing tests further up in my terminal 😞

**Comment by kraih** [comment]
There's probably a smart way to generate better Perl with `Mojo::Template`.

**Comment by marcusramberg** [comment]
I'm gonna go ahead and close this, since it seems to have gone stale, and will not be approved.
